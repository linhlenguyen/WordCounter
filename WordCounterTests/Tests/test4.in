
lndevnotes
ARCHIVE
Reader Monad
Don’t be scared! The reader monad is actually not so complicated, and has real easy-to-use utility.
There are two ways of approaching a monad: we can ask
What does the monad do? What operations is it equipped with? What is it good for?
How is the monad implemented? From where does it arise?
From the first approach, the reader monad is some abstract type
data Reader env a
such that
-- Reader is a monad
instance Monad (Reader env)

-- and we have a function to get its environment
ask :: Reader env env

-- finally, we can run a Reader
runReader :: Reader env a -> env -> a
So how do we use this? Well, the reader monad is good for passing (implicit) configuration information through a computation.
Any time you have a “constant” in a computation that you need at various points, but really you would like to be able to perform the same computation with different values, then you should use a reader monad.
Reader monads are also used to do what the OO people call dependency injection. For example, the negamax algorithm is used frequently (in highly optimized forms) to compute the value of a position in a two player game. The algorithm itself though does not care what game you are playing, except that you need to be able to determine what the “next” positions are in the game, and you need to be able to tell if the current position is a victory position.
 import Control.Monad.Reader

 data GameState = NotOver | FirstPlayerWin | SecondPlayerWin | Tie

 data Game position
   = Game {
           getNext :: position -> [position],
           getState :: position -> GameState
          }

 getNext' :: position -> Reader (Game position) [position]
 getNext' position
   = do game <- ask
        return $ getNext game position

 getState' :: position -> Reader (Game position) GameState
 getState' position
   = do game <- ask
        return $ getState game position


 negamax :: Double -> position -> Reader (Game position) Double
 negamax color position
     = do state <- getState' position
          case state of
             FirstPlayerWin -> return color
             SecondPlayerWin -> return $ negate color
             Tie -> return 0
             NotOver -> do possible <- getNext' position
                           values <- mapM ((liftM negate) . negamax (negate color)) possible
                           return $ maximum values
This will then work with any finite, deterministic, two player game.
This pattern is useful even for things that are not really dependency injection. Suppose you work in finance, you might design some complicated logic for pricing an asset (a derivative say), which is all well and good and you can do without any stinking monads. But then, you modify your program to deal with multiple currencies. You need to be able to convert between currencies at the fly. Your first attempt is to define a top level function
type CurrencyDict = Map CurrencyName Dollars
currencyDict :: CurrencyDict
to get spot prices. You can then call this dictionary in your code….but wait! That won’t work! The currency dictionary is immutable and so has to be the same not only for the life of your program, but from the time it gets compiled! So what do you do? Well one option would be to use the Reader monad:
 computePrice :: Reader CurrencyDict Dollars
 computePrice
    = do currencyDict <- ask
         --insert computation here
Perhaps the most classic use-case is in implementing interpreters. But, before we look at that, we need to introduce another function
 local :: (env -> env) -> Reader env a -> Reader env a
Okay, so Haskell and other functional languages are based on the lambda calculus. Lambda calculus has a syntax that looks like
 data Term = Apply Term Term | Lambda String Term | Var Term deriving (Show)
and we want to write an evaluator for this language. To do so, we will need to keep track of an environment, which is a list of bindings associated with terms (actually it will be closures because we want to do static scoping).
 newtype Env = Env ([(String,Closure)])
 type Closure = (Term,Env)
When we are done we should get out a value (or an error):
 data Value = Lam String Closure | Failure String
So, lets write the interpreter:
interp' :: Term -> Reader Env Value
--when we have lambda term, we can just return it
interp' (Lambda nv t)
   = do env <- ask
        return $ Lam nv (t,env)
--when we run into a value we look it up in the environment
interp' (Var v)
   = do (Env env) <- ask
        case lookup (show v) env of
          -- if it is not in the environment we have a problem
          Nothing -> return . Failure $ "unbound variable: " ++ (show v)
          -- if it is in the environment, than we should interpret it
          Just (term,env) -> local (const env) $ interp' term
--the complicated case is an application
interp' (Apply t1 t2)
   = do v1 <- interp' t1
        case v1 of
           Failure s -> return (Failure s)
           Lam nv clos -> local (\(Env ls) -> Env ((nv,clos):ls)) $ interp' t2
--I guess not that complicated!
Finally, we can use it by passing a trivial environment:
interp :: Term -> Value
interp term = runReader (interp' term) (Env [])
And that is it. A fully functional interpreter for the lambda calculus.
So, the other way to think about this is to ask: how is it implemented? Well the answer is that the reader monad is actually one of the simplest and most elegant of all monads.
newtype Reader env a = Reader {runReader :: env -> a}
Reader is just a fancy name for functions! We have already defined runReader so what about the other parts of the API? Well every Monad is also a Functor:
instance Functor (Reader env) where
   fmap f (Reader g) = Reader $ f . g
Now, to get a monad:
instance Monad (Reader env) where
   return x = Reader (\_ -> x)
   (Reader f) >>= g = Reader $ \x -> runReader (g (f x)) x
which is not so scary. ask is really simple:
ask = Reader $ \x -> x
while local isn’t so bad.
local f (Reader g) = Reader $ \x -> runReader g (f x)
Okay, so the reader monad is just a function. Why have Reader at all? Good question. Actually, you don’t need it!
instance Functor ((->) env) where
   fmap = (.)

 instance Monad ((->) env) where
   return = const
   f >>= g = \x -> g (f x) x
These are even simpler. What is more, ask is just id and local is just function composition in the other order!
http://stackoverflow.com/questions/14178889/what-is-the-purpose-of-the-reader-monad



The State Monad: A Tutorial for the Confused?
http://brandon.si/code/the-state-monad-a-tutorial-for-the-confused/
I’ve written this brief tutorial on haskell’s State monad to help bridge some of the elusive gaps that I encountered in other explanations I’ve read, and to try to cut through all the sticky abstraction. This is written for someone who has a good understanding of the Maybe and Listmonads, but has gotten stuck trying to understand State. I hope it’s helpful!
The Data Declaration:

To understand a monad you look at it’s datatype and then at the definition for bind (>>=). Most monad tutorials start by showing you the data declaration of a State s a in passing, as if it needed no explanation:
newtype State s a = State { runState :: s -> (a, s) }
But this does need explanation! This is crazy stuff and nothing like what we’ve seen before in the list monad or the Maybe monad
The constructor State holds a function, not just a simple value like Maybe’s Just. This looks weird.
Furthermore there is an accessor function runState with a weirdly imperative-sounding name.
Finally, there are two free variables on the left side, not just one.
Yikes! Let’s try to get our head on straight and figure this out:
First of all the State monad is just an abstraction for a function that takes a state and returns an intermediate value and some new state value. To formalize this abstraction in haskell, we wrap the function in the newtype State allowing us to define a Monad class instance.
Stepping back from the abstract and conceptual, what we have is the State constructor acting as a container for a function :: s -> (a,s), while the definition for bind just provides a mechanism for “composing” a function state -> (val,state) within the State wrapper.
Just as you can chain together functions using (.) as in (+1) . (*3) . head :: (Num a) => [a] -> a, the state monad gives you (>>=) to chain together functions that look essentially like :: a -> s -> (a,s)into a single function :: s -> (a,s).
Let’s bring the discussion back to actual code and try to make sure we understand those three points of weirdness outlined above. Here’s a stupid example of a function that can be “contained” in our state type:
-- look at our counter and return "foo" or "bar"
-- along with the incremented counter:
fromStoAandS :: Int -> (String,Int)
fromStoAandS c | c `mod` 5 == 0 = ("foo",c+1)
               | otherwise = ("bar",c+1)
If we just wrap that in a State constructor, we’re in the State monad:
stateIntString :: State Int String
stateIntString = State fromStoAandS
But what about runState? All that does of course is give us the “contents” of our State constructor: i.e. a single function :: s -> (a,s). It could have been named stateFunction but someone thought it would be really clever to be able to write things like:
runState stateIntString 1
See, all we’ve done there is used runState to take our function (fromStoAandS) out of the State wrapper; it is then applied to its initial state (1). We would do this runState business after building up our composed function with (>>=), mapM, etc.
That leaves point 3 unanswered. Let’s start exploring the instance declaration for State.
The Instance Declaration

We’ll start with the first line:
instance Monad (State s) where
We create a Monad instance for (State s) not State. You can think of this as a partially-applied type, which is equivalent to a partially-applied function:
(State)     <==> (+)
(State s)   <==> (1+)
(State s a) <==> (1+2)
So (State s) is the m in our m a. This means the type of our state will remain the same as we compose our function with (>>=), whereas the intermediate values (the as) may well change type as they move through the chain.
Before we move on to the meat of the instance declaration, I’d like to get your mind calibrated to look at the definitions for return and (>>=):
Whenever you see m a, as in
return :: (Monad m) => a -> m a
…remember that m a is actually
State s a
…and when you remember (State s a), think
(s -> (s,a))
So in your mind, m a becomes function :: s -> (a,s) everywhere you see it. Just forget about the silly State wrapper (the compiler does)!
The definition of return and Bind

Let’s wet our feet with the definition for return:
return a = State $ \s -> (a, s)
All return does is take some value a and make a function that takes a state value and returns (value, state value). If we ignore the whole State wrapping business, then return is just (,) :: a -> b -> (a, b)
Now recall the definition of bind:
(>>=) :: (Monad m) => m a -> (a -> m b) -> m b
Which in our case is:
(>>=) :: State s a        ->
         (a -> State s b) ->
         State s b
And which is just a silly abstraction for the super special function composition that’s going on, which looks like:
(>>=) :: (s -> (a,s))      ->
         (a -> s -> (b,s)) ->
         (s -> (b,s))
So on the left hand side of (>>=) is a function that takes some initial state and produces a (value,new_state). On the right hand side is a function that takes that value and that new_state and generates it’s own (new_value, newer_state). The job of bind is simply to combine those two functions into one bigger function from the initial state to (new_value,newer_state), just like the simple function composition operator (.) :: (b -> c) -> (a -> b) -> a -> c
At this point, we can show you bind’s definition:
m >>= k = State $ \s -> let (a, s') = runState m s
                         in runState (k a) s'
You can work through that on your own, keeping in mind that we’re doing function composition here. The main thing to remember is that the s at the top, right after State, won’t actually be bound to a value until we unwrap the function with runState and pass it the initial state value, at which point we can evaluate the entire chain.
A Final Note About The State Monad with do Notation

State is often used like this:
stateFunction :: State [a] ()
stateFunction = do x <- pop
                   pop
                   push x
Remember that the functions above are desugaring to m >>= \a-> f...or if there is no left arrow on the previous line: m >>= \_-> f... That a in there is an intermediate value, the fstin the tuple. The push function might look like:
push :: State [a] ()
push a = State $ \as -> (() , a:as)
The function doesn’t return any meaningful a value, so we don’t bind it by using the <-. For more work with do notation and some fine pictures, see Bonus’s post on something awful.
Getting more general: StateT and MonadState

added by request 2/16/2012
So now you’re an expert on the State monad. Unfortunately (actually it’s a good thing) the State type I describe above isn’t in any of the standard libraries. Instead State is defined in terms of the StateT monad transformer here.
type State s = StateT s Identity
If you haven’t seen Monad transformers before, see if you can figure out how StateT s Identity is equivalent to State s as I defined it above. Just follow the links on hackage.
You might also have noticed a typeclass called MonadState, also in the mtl package, and be wondering how that fits in. Here’s what it looks like; I’ll explain the odd-looking bits in a moment:
class Monad m => MonadState s m | m -> s where
    get :: m s
    put :: s -> m ()
Whereas above we were discussing State, a concrete data type, MonadState is a new typeclassfor types that are monads and for which we can define get and put operations.
The class allows for a variety of Monad Transformer “stacks” that use state-passing to share a common interface for the basic state operations.
Small aside: since I neglected doing this above, this is how we would define get and put as regular functions (not methods of MonadState) on our State type:
-- return the state value being passed around:
get :: State s s
get = State $ \s -> (s,s)

-- replace the current state value with 's':
put :: s -> State s ()
put s = State $ \_ -> (s,())
Quite useful. See if you can understand how those work now that you’ve got a better grasp of State. Excercise: define modify :: (s->s) -> State s ().
Back to MonadState: If this class and its instances look confusing to you, you need to know about two extensions to haskell-98 (both of which are very common and sticking around): Multi-Parameter Type Classes, and Functional Dependencies.
In GHC you can enable both these extensions in your source by putting this at the top:
{-# LANGUAGE MultiParamTypeClasses, FunctionalDependencies #-}
Multi-parameter Type Classes
Ignoring the functional dependencies (a.k.a “fundeps”) in MonadState you get:
class Monad m => MonadState s m where
In which…
MonadState is the name of the class
s is the first class type variable (in this case the type of our state)
m is the second type variable (the type of our state-like monad, e.g. State s)
Multi-parameter type classes are tricky in that they define a relationship between multiple typeswith associated operations. You can read more about them in the GHC docs.
Functional dependencies
With multiple parameters in a single class, you can often end up with instances that are disallowed or difficult to use because they are ambiguous. Functional Dependencies help resolve ambiguity by allowing a way to specify that certain type parameters can be determined by knowledge of one or more of the other parameters.
In the case of MonadState, the part of the class declaration that looks like:
| m -> s
says that the type of m (say StateT Int IO) uniquely determines the type of s (Int). Again please see the linked GHC docs for details; the Collects example used is very similar to MonadState.
Putting it all together, the class declaration might read in english as:
The relation of types m and s where m uniquely determines s is in the MonadState class.
MonadState
instance for State
Finally here’s what the MonadState instance for the State type we’ve been discussing would look like (but again, it’s not because mtl builds around the more general StateT):
instance MonadState s (State s) where
    get = State $ \s -> (s,s)
    put s = State $ \_ -> (s,())



Haskell: How is <*> pronounced?
A great answer from Stack overflow
As you’re probably aware, Haskell borrows a few bits of terminology from various fields of abstract math, most notably Category Theory, from whence we get functors and monads. The use of these terms in Haskell diverges somewhat from the formal mathematical definitions, but they’re usually close enough to be good descriptive terms anyway.
The Applicative type class sits somewhere between Functor and Monad, so one would expect it to have a similar mathematical basis. The documentation for the Control.Applicative module begins with:
This module describes a structure intermediate between a functor and a monad: it provides pure expressions and sequencing, but no binding. (Technically, a strong lax monoidal functor.)
Hmm.
class (Functor f) => StrongLaxMonoidalFunctor f where
    . . .
Not quite as catchy as Monad, I think.
What all this basically boils down to is that Applicative doesn’t correspond to any concept that’s particularly interesting mathematically, so there’s no ready-made terms lying around that capture the way it’s used in Haskell. So, set the math aside for now.
If we want to know what to call (<*>) it might help to know what it basically means.
So what’s up with Applicative, anyway, and why do we call it that?
What Applicative amounts to in practice is a way to lift arbitrary functions into a Functor. Consider the combination of Maybe (arguably the simplest non-trivial Functor) and Bool (likewise the simplest non-trivial data type).
maybeNot :: Maybe Bool -> Maybe Bool
maybeNot = fmap not
The function fmap lets us lift not from working on Bool to working on Maybe Bool. But what if we want to lift (&&)?
maybeAnd' :: Maybe Bool -> Maybe (Bool -> Bool)
maybeAnd' = fmap (&&)
Well, that’s not what we want at all! In fact, it’s pretty much useless. We can try to be clever and sneak another Bool into Maybe through the back…
maybeAnd'' :: Maybe Bool -> Bool -> Maybe Bool
maybeAnd'' x y = fmap ($ y) (fmap (&&) x)
…but that’s no good. For one thing, it’s wrong. For another thing, it’s ugly. We could keep trying, but it turns out that there’s no way to lift a function of multiple arguments to work on an arbitrary Functor. Annoying!
On the other hand, we could do it easily if we used Maybe’s Monad instance:
maybeAnd :: Maybe Bool -> Maybe Bool -> Maybe Bool
maybeAnd x y = do x' <- x
                  y' <- y
                  return (x' && y')
Now, that’s a lot of hassle just to translate a simple function–which is why Control.Monad provides a function to do it automatically, liftM2. The 2 in its name refers to the fact that it works on functions of exactly two arguments; similar functions exist for 3, 4, and 5 argument functions. These functions arebetter, but not perfect, and specifying the number of arguments is ugly and clumsy.
Which brings us to the paper that introduced the Applicative type class. In it, the authors make essentially two observations:
Lifting multi-argument functions into a Functor is a very natural thing to do
Doing so doesn’t require the full capabilities of a Monad
Normal function application is written by simple juxtaposition of terms, so to make “lifted application” as simple and natural as possible, the paper introduces infix operators to stand in for application, lifted into the Functor, and a type class to provide what’s needed for that.
All of which brings us to the following point: (<*>) simply represents function application–so why pronounce it any differently than you do the whitespace “juxtaposition operator”?
But if that’s not very satisfying, we can observe that the Control.Monad module also provides a function that does the same thing for monads:
ap :: (Monad m) => m (a -> b) -> m a -> m b
Where ap is, of course, short for “apply”. Since any Monad can be Applicative, and ap needs only the subset of features present in the latter, we can perhaps say that if (<*>) weren’t an operator, it should be called ap.
We can also approach things from the other direction. The Functor lifting operation is called fmapbecause it’s a generalization of the map operation on lists. What sort of function on lists would work like (<*>)? There’s what ap does on lists, of course, but that’s not particularly useful on its own.
In fact, there’s a perhaps more natural interpretation for lists. What comes to mind when you look at the following type signature?
listApply :: [a -> b] -> [a] -> [b]
There’s something just so tempting about the idea of lining the lists up in parallel, applying each function in the first to the corresponding element of the second. Unfortunately for our old friend Monad, this simple operation violates the monad laws if the lists are of different lengths. But it makes a fine Applicative, in which case (<*>) becomes a way of stringing together a generalized version ofzipWith, so perhaps we can imagine calling it fzipWith?
This zipping idea actually brings us full circle. Recall that math stuff earlier, about monoidal functors? As the name suggests, these are a way of combining the structure of monoids and functors, both of which are familiar Haskell type classes:
class Functor f where
    fmap :: (a -> b) -> f a -> f b

class Monoid a where
    mempty :: a
    mappend :: a -> a -> a
What would these look like if you put them in a box together and shook it up a bit? From Functorwe’ll keep the idea of a structure independent of its type parameter, and from Monoid we’ll keep the overall form of the functions:
class (Functor f) => MonoidalFunctor f where
    mfEmpty :: f ?
    mfAppend :: f ? -> f ? -> f ?
We don’t want to assume that there’s a way to create an truly “empty” Functor, and we can’t conjure up a value of an arbitrary type, so we’ll fix the type of mfEmpty as f ().
We also don’t want to force mfAppend to need a consistent type parameter, so now we have this:
class (Functor f) => MonoidalFunctor f where
    mfEmpty :: f ()
    mfAppend :: f a -> f b -> f ?
What’s the result type for mfAppend? We have two arbitrary types we know nothing about, so we don’t have many options. The most sensible thing is to just keep both:
class (Functor f) => MonoidalFunctor f where
    mfEmpty :: f ()
    mfAppend :: f a -> f b -> f (a, b)
At which point mfAppend is now clearly a generalized version of zip on lists, and we can reconstruct Applicative easily:
mfPure x = fmap (\() -> x) mfEmpty
mfApply f x = fmap (\(f, x) -> f x) (mfAppend f x)
This also shows us that pure is related to the identity element of a Monoid, so other good names for it might be anything suggesting a unit value, a null operation, or such.
That was lengthy, so to summarize:
(<*>) is just a modified function application, so you can either read it as “ap” or “apply”, or elide it entirely the way you would normal function application.
(<*>) also roughly generalizes zipWith on lists, so you can read it as “zip functors with”, similarly to reading fmap as “map a functor with”.
The first is closer to the intent of the Applicative type class–as the name suggests–so that’s what I recommend.
In fact, I encourage liberal use, and non-pronunciation, of all lifted application operators:
(<$>), which lifts a single-argument function into a Functor
(<*>), which chains a multi-argument function through an Applicative
(=<<), which binds a function that enters a Monad onto an existing computation
All three are, at heart, just regular function application, spiced up a little bit.



 
