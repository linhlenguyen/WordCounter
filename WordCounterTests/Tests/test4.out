a, 171
the, 167
s, 115
to, 89
is, 77
of, 76
State, 59
and, 58
that, 56
we, 54
it, 52
in, 50
function, 44
f, 38
you, 38
for, 38
type, 32
Reader, 31
m, 31
can, 29
monad, 27
Monad, 27
x, 26
do, 26
state, 26
so, 24
as, 23
The, 23
on, 23
return, 23
env, 22
value, 22
t, 22
position, 21
class, 21
this, 20
with, 20
b, 19
what, 19
be, 19
are, 19
not, 18
which, 18
just, 18
Functor, 17
like, 17
where, 16
Maybe, 16
an, 16
MonadState, 15
our, 15
Bool, 15
get, 14
at, 14
instance, 14
I, 13
if, 13
functions, 13
have, 13
but, 12
look, 11
ask, 11
fmap, 11
Applicative, 11
We, 10
would, 10
from, 10
need, 10
What, 10
interp, 10
So, 10
use, 10
all, 10
about, 9
also, 9
way, 9
two, 9
runState, 9
there, 9
does, 9
g, 9
This, 9
other, 9
or, 9
application, 9
case, 9
Term, 8
into, 8
up, 8
good, 8
by, 8
lists, 8
work, 8
data, 8
game, 8
Env, 8
want, 7
one, 7
any, 7
But, 7
thing, 7
your, 7
mfAppend, 7
used, 7
In, 7
more, 7
reader, 7
then, 7
actually, 7
us, 7
put, 7
simple, 7
some, 7
name, 6
first, 6
maybeAnd, 6
above, 6
c, 6
its, 6
terms, 6
ap, 6
no, 6
might, 6
ve, 6
environment, 6
bind, 6
definition, 6
Haskell, 6
local, 6
You, 6
define, 6
String, 6
y, 6
monads, 6
StateT, 6
how, 6
new, 6
through, 6
runReader, 6
If, 5
re, 5
together, 5
looks, 5
computation, 5
declaration, 5
read, 5
able, 5
has, 5
now, 5
see, 5
ll, 5
mfEmpty, 5
point, 5
Just, 5
keep, 5
don, 5
getNext, 5
out, 5
term, 5
Well, 5
Int, 5
Game, 5
getState, 5
v, 5
understand, 4
initial, 4
provides, 4
here, 4
think, 4
most, 4
arbitrary, 4
Failure, 4
That, 4
A, 4
lift, 4
fromStoAandS, 4
chain, 4
abstraction, 4
when, 4
intermediate, 4
try, 4
nv, 4
composition, 4
All, 4
operation, 4
argument, 4
fact, 4
takes, 4
make, 4
constructor, 4
single, 4
Let, 4
newtype, 4
these, 4
same, 4
call, 4
should, 4
values, 4
really, 4
around, 4
passing, 4
multiple, 4
mind, 4
negamax, 4
parameter, 4
push, 4
know, 4
between, 4
functors, 4
Control, 4
complicated, 4
color, 4
both, 4
they, 4
For, 4
very, 4
operations, 4
will, 4
doesn, 3
called, 3
negate, 3
trivial, 3
GameState, 3
similar, 3
Value, 3
Finally, 3
wrapper, 3
those, 3
k, 3
arguments, 3
variable, 3
show, 3
idea, 3
Lam, 3
math, 3
back, 3
GHC, 3
Functional, 3
There, 3
start, 3
natural, 3
Multi, 3
help, 3
haskell, 3
written, 3
From, 3
brings, 3
left, 3
abstract, 3
perhaps, 3
side, 3
parameters, 3
Now, 3
types, 3
Closure, 3
defined, 3
simplest, 3
example, 3
possible, 3
lifted, 3
things, 3
top, 3
calculus, 3
why, 3
MonoidalFunctor, 3
functional, 3
stateFunction, 3
useful, 3
say, 3
course, 3
lambda, 3
functor, 3
stateIntString, 3
hand, 3
applied, 3
Monoid, 3
To, 3
write, 3
code, 3
because, 3
before, 3
another, 3
CurrencyDict, 3
module, 3
remember, 3
Lambda, 3
own, 3
structure, 3
operator, 3
could, 3
non, 3
second, 2
definitions, 2
dependencies, 2
classes, 2
See, 2
paper, 2
take, 2
head, 2
basically, 2
clever, 2
counter, 2
seen, 2
equivalent, 2
Dependencies, 2
juxtaposition, 2
Declaration, 2
full, 2
partially, 2
create, 2
Classes, 2
Type, 2
line, 2
needed, 2
after, 2
business, 2
trying, 2
extensions, 2
explanation, 2
move, 2
multi, 2
figure, 2
stuff, 2
nothing, 2
instances, 2
regular, 2
becomes, 2
returns, 2
using, 2
maybeNot, 2
right, 2
essentially, 2
means, 2
ugly, 2
working, 2
pop, 2
often, 2
mathematical, 2
anyway, 2
doing, 2
newer, 2
simply, 2
bar, 2
three, 2
At, 2
foo, 2
Here, 2
fine, 2
them, 2
something, 2
As, 2
docs, 2
aside, 2
e, 2
common, 2
discussing, 2
silly, 2
wrap, 2
allowing, 2
bits, 2
uniquely, 2
determines, 2
monoidal, 2
mtl, 2
Which, 2
pure, 2
easily, 2
On, 2
Identity, 2
Unfortunately, 2
general, 2
particularly, 2
been, 2
points, 2
injection, 2
time, 2
computePrice, 2
associated, 2
various, 2
isn, 2
while, 2
map, 2
current, 2
even, 2
only, 2
Okay, 2
answer, 2
Dollars, 2
currencyDict, 2
done, 2
bit, 2
different, 2
dictionary, 2
And, 2
ls, 2
clos, 2
const, 2
element, 2
than, 2
lets, 2
dependency, 2
won, 2
interpreter, 2
list, 2
currencies, 2
These, 2
suggests, 2
operators, 2
player, 2
zip, 2
someone, 2
following, 2
Apply, 2
liftM, 2
mapM, 2
Double, 2
Tie, 2
Var, 2
How, 2
tutorial, 2
NotOver, 2
apply, 2
implemented, 2
generalized, 2
program, 2
http, 2
modify, 2
run, 2
well, 2
FirstPlayerWin, 2
algorithm, 2
version, 2
approach, 2
SecondPlayerWin, 2
such, 2
suggesting, 1
few, 1
aware, 1
terminology, 1
notably, 1
was, 1
fields, 1
lengthy, 1
unit, 1
null, 1
borrows, 1
infix, 1
Category, 1
clearly, 1
reconstruct, 1
describes, 1
mfPure, 1
begins, 1
documentation, 1
basis, 1
expect, 1
somewhere, 1
sits, 1
mfApply, 1
anything, 1
descriptive, 1
close, 1
usually, 1
shows, 1
formal, 1
related, 1
identity, 1
somewhat, 1
diverges, 1
names, 1
whence, 1
Theory, 1
enough, 1
probably, 1
Putting, 1
modified, 1
end, 1
zipWith, 1
With, 1
similarly, 1
reading, 1
closer, 1
typeswith, 1
relationship, 1
tricky, 1
intent, 1
recommend, 1
fundeps, 1
Ignoring, 1
FunctionalDependencies, 1
encourage, 1
MultiParamTypeClasses, 1
liberal, 1
pronunciation, 1
LANGUAGE, 1
lifts, 1
putting, 1
chains, 1
binds, 1
enters, 1
onto, 1
existing, 1
heart, 1
disallowed, 1
summarize, 1
difficult, 1
resolve, 1
overflow, 1
either, 1
Stack, 1
great, 1
pronounced, 1
builds, 1
again, 1
elide, 1
relation, 1
english, 1
entirely, 1
Collects, 1
details, 1
linked, 1
please, 1
Again, 1
normal, 1
IO, 1
says, 1
roughly, 1
part, 1
knowledge, 1
determined, 1
generalizes, 1
certain, 1
specify, 1
ambiguity, 1
ambiguous, 1
sensible, 1
empty, 1
many, 1
fmapbecause, 1
generalization, 1
sort, 1
exist, 1
interpretation, 1
comes, 1
signature, 1
listApply, 1
exactly, 1
works, 1
refers, 1
automatically, 1
tempting, 1
translate, 1
hassle, 1
lot, 1
lining, 1
Annoying, 1
parallel, 1
applying, 1
turns, 1
each, 1
wrong, 1
corresponding, 1
sneak, 1
useless, 1
old, 1
arebetter, 1
lifting, 1
source, 1
perfect, 1
Normal, 1
capabilities, 1
stand, 1
provide, 1
require, 1
Doing, 1
represents, 1
pronounce, 1
differently, 1
whitespace, 1
satisfying, 1
observe, 1
Where, 1
much, 1
short, 1
needs, 1
subset, 1
Lifting, 1
observations, 1
authors, 1
features, 1
introduced, 1
present, 1
clumsy, 1
number, 1
specifying, 1
latter, 1
weren, 1
Since, 1
options, 1
pretty, 1
violates, 1
mathematically, 1
interesting, 1
overall, 1
form, 1
assume, 1
concept, 1
correspond, 1
down, 1
boils, 1
truly, 1
introduces, 1
conjure, 1
catchy, 1
quite, 1
Not, 1
StrongLaxMonoidalFunctor, 1
Hmm, 1
fix, 1
lax, 1
strong, 1
Technically, 1
force, 1
binding, 1
consistent, 1
sequencing, 1
expressions, 1
result, 1
ready, 1
made, 1
independent, 1
Functorwe, 1
laws, 1
lengths, 1
makes, 1
stringing, 1
ofzipWith, 1
imagine, 1
calling, 1
likewise, 1
fzipWith, 1
zipping, 1
arguably, 1
circle, 1
combination, 1
friend, 1
Consider, 1
earlier, 1
combining, 1
monoids, 1
familiar, 1
mempty, 1
mappend, 1
practice, 1
amounts, 1
box, 1
set, 1
capture, 1
lying, 1
shook, 1
Recall, 1
direction, 1
lndevnotes, 1
Parameter, 1
fully, 1
elegant, 1
fancy, 1
already, 1
parts, 1
API, 1
guess, 1
every, 1
bad, 1
Why, 1
Good, 1
question, 1
Actually, 1
simpler, 1
scary, 1
id, 1
interpret, 1
Nothing, 1
syntax, 1
deriving, 1
Show, 1
evaluator, 1
language, 1
track, 1
unbound, 1
bindings, 1
static, 1
scoping, 1
When, 1
error, 1
lookup, 1
problem, 1
closures, 1
order, 1
stackoverflow, 1
com, 1
stuck, 1
hope, 1
helpful, 1
Data, 1
datatype, 1
Most, 1
gotten, 1
tutorials, 1
crazy, 1
holds, 1
weird, 1
Furthermore, 1
accessor, 1
weirdly, 1
showing, 1
Listmonads, 1
understanding, 1
who, 1
questions, 1
purpose, 1
Tutorial, 1
Confused, 1
brandon, 1
si, 1
confused, 1
brief, 1
bridge, 1
elusive, 1
gaps, 1
encountered, 1
explanations, 1
cut, 1
sticky, 1
based, 1
imperative, 1
languages, 1
interpreters, 1
people, 1
frequently, 1
highly, 1
optimized, 1
forms, 1
compute, 1
OO, 1
itself, 1
care, 1
playing, 1
except, 1
determine, 1
next, 1
positions, 1
though, 1
tell, 1
perform, 1
Any, 1
ARCHIVE, 1
Don, 1
scared, 1
real, 1
easy, 1
utility, 1
constant, 1
ways, 1
equipped, 1
arise, 1
finally, 1
implicit, 1
configuration, 1
information, 1
approaching, 1
victory, 1
import, 1
maximum, 1
CurrencyName, 1
spot, 1
prices, 1
wait, 1
currency, 1
immutable, 1
Map, 1
life, 1
compiled, 1
option, 1
insert, 1
Perhaps, 1
classic, 1
implementing, 1
gets, 1
level, 1
attempt, 1
Your, 1
finite, 1
deterministic, 1
pattern, 1
Suppose, 1
finance, 1
design, 1
logic, 1
pricing, 1
asset, 1
derivative, 1
without, 1
stinking, 1
deal, 1
convert, 1
fly, 1
introduce, 1
enable, 1
sounding, 1
variables, 1
post, 1
awful, 1
Getting, 1
added, 1
request, 1
expert, 1
Bonus, 1
describe, 1
libraries, 1
Instead, 1
transformer, 1
haven, 1
transformers, 1
follow, 1
standard, 1
links, 1
pictures, 1
meaningful, 1
unwrap, 1
pass, 1
evaluate, 1
entire, 1
Final, 1
Note, 1
notation, 1
About, 1
Remember, 1
desugaring, 1
arrow, 1
previous, 1
fstin, 1
tuple, 1
Notation, 1
hackage, 1
noticed, 1
typeclass, 1
since, 1
neglected, 1
methods, 1
being, 1
passed, 1
replace, 1
Small, 1
Quite, 1
better, 1
grasp, 1
Excercise, 1
Back, 1
confusing, 1
sticking, 1
got, 1
basic, 1
interface, 1
share, 1
package, 1
wondering, 1
fits, 1
explain, 1
odd, 1
looking, 1
moment, 1
Whereas, 1
were, 1
concrete, 1
typeclassfor, 1
allows, 1
variety, 1
Transformer, 1
stacks, 1
until, 1
free, 1
bound, 1
keeping, 1
weirdness, 1
outlined, 1
stupid, 1
contained, 1
along, 1
incremented, 1
sure, 1
mod, 1
give, 1
contents, 1
i, 1
It, 1
spiced, 1
named, 1
otherwise, 1
thought, 1
actual, 1
bring, 1
Yikes, 1
straight, 1
First, 1
formalize, 1
Stepping, 1
conceptual, 1
discussion, 1
acting, 1
mechanism, 1
composing, 1
val, 1
within, 1
Num, 1
gives, 1
container, 1
building, 1
composed, 1
etc, 1
feet, 1
ignore, 1
whole, 1
wrapping, 1
recall, 1
super, 1
wet, 1
special, 1
produces, 1
generates, 1
job, 1
combine, 1
bigger, 1
let, 1
going, 1
Bind, 1
compiler, 1
forget, 1
leaves, 1
unanswered, 1
exploring, 1
Instance, 1
remain, 1
compose, 1
whereas, 1
may, 1
change, 1
Before, 1
meat, 1
d, 1
calibrated, 1
Whenever, 1
everywhere, 1
main, 1
little, 1

